app/src/org/commcare/android/tasks/TemplatePrinterTask.java

```diff
@@ -0,0 +1,253 @@
+     * with attribute values.
+     * @param input String input
+     * @param values Bundle of String attribute key-value mappings
+     * @return The populated String
+     */
+    private static String replace(String input, Bundle values) {
+        // Split input into tokens bounded by {{ and }}
+        String[] tokens = TemplatePrinterUtils.splitKeepDelimiter(input, "\\{{2}", "\\}{2}");
+        for (int i=0; i<tokens.length; i++) {
+            String token = tokens[i];
+
+            // Every 2nd token is a attribute enclosed in {{ }}
+            if (i % 2 == 1) {
+
+                // Split token into tokenSplits bounded by < and >
+                String[] tokenSplits = TemplatePrinterUtils.splitKeepDelimiter(token, "<|(\\}{2})", ">|(\\{{2})");
+
+                // First and last tokenSplits are {{ and }}
+                for (int j=1; j<tokenSplits.length-1; j++) {
+                    String tokenSplit = tokenSplits[j];
+
+                    // tokenSplit is key or whitespace
+                    if (!tokenSplit.startsWith("<")) {
+
+                        // Remove whitespace from key
+                        String key = TemplatePrinterUtils.remove(tokenSplit, " ");
+
+                        if (values.containsKey(key) && (key = values.getString(key)) != null) {
+                            // Populate with value
+                            tokenSplits[j] = key;
+                        } else {
+                            // Empty if not found
+                            tokenSplits[j] = "";
+                        }
+                    }
+
+                }
+
+                // Remove {{ and }}
+                tokenSplits[0] = "";
+                tokenSplits[tokenSplits.length-1] = "";
+
+                // Reconstruct token
+                tokens[i] = TemplatePrinterUtils.join(tokenSplits);
+            }
+        }
+
+        // Reconstruct input
+        return TemplatePrinterUtils.join(tokens);
+    }
+
+    /**
+     * Validates the input string for well-formed {{ }} and < > pairs.
+     * If malformed, throws a RuntimeException which will be caught by
+     * doInBackground(), and trigger the appropriate result code to be
+     * sent back to the attached PopulateListener
+     * 
+     * @param input String to validate
+     */
+    private static void validateStringOrThrowException(String input) {
+
+        boolean isBetweenMustaches = false;
+        boolean isBetweenChevrons = false;
+        StringBuilder recentString = new StringBuilder();
+
+        for (int i = 0; i < input.length(); i++) {
+            char c = input.charAt(i);
+            recentString.append(c);
+
+            if (recentString.length() > 40) {
+                recentString.deleteCharAt(0);
+            }
+
+            if (c == '{') {
+                if (isBetweenMustaches) {
+                    validationErrorType = VALIDATION_ERROR_MUSTACHE;
+                    throw new RuntimeException(recentString.toString());
```

Comment:
You should create an explicit exception class that you then raise here and catch it in doInBackground etc. It is bad to raise a generic error, because other code might also produce such errors, so when you catch them later you have no idea where the exception came from.